chain/jsonrpc/src/lib.rs

```diff
@@ -243,11 +244,32 @@ impl JsonRpcHandler {
         let timer = Instant::now();
 
         let request_method = request.method.clone();
+        let params: RpcQueryRequest = RpcRequest::parse(request.params.clone())?;
         let response = self.process_request_internal(request).await;
 
         let request_method = match &response {
             Err(err) if err.code == -32_601 => "UNSUPPORTED_METHOD",
-            _ => &request_method,
+            _ => {
+                // `query` RPC method consists of several methods which should be tracked separately
+                if request_method == "query" {
+                    match params.request {
+                        QueryRequest::ViewAccount { .. } => "query_view_account",
+                        QueryRequest::ViewCode { .. } => "query_view_code",
+                        QueryRequest::ViewState { include_proof, .. } => {
+                            if include_proof {
+                                "query_view_state_with_proof"
+                            } else {
+                                "query_view_state"
+                            }
+                        }
+                        QueryRequest::ViewAccessKey { .. } => "query_view_access_key",
+                        QueryRequest::ViewAccessKeyList { .. } => "query_view_access_key_list",
+                        QueryRequest::CallFunction { .. } => "query_call_function",
+                    }
+                } else {
+                    &request_method
+                }
+            }
         };
 
         metrics::HTTP_RPC_REQUEST_COUNT.with_label_values(&[request_method]).inc();
```

Comment:
`request_method` should be determined based on the request and not on the response.
Also, the number of requests should be counted the request is processed.

While you're fixing the `query` instrumentation, do you mind also fixing that?

Reply:
Context: Counting of requests after processing and not before processing is done to limit the set of method names retained in metrics.
If somebody makes an RPC with a random string as method name, that method name will be kept in memory in Prometheus metrics until the node restart.

Should be fine to keep the current behavior, as the alternative is worse, and I don't see a clean way to safely count requests before processing.

Reply:
@khorolets @nikurt I've created a little other solution.
When it's not `method_name`, but `metrics_name`, it looks logical that `metrics_name` may depend on the response.
I create `metrics_name` inside `process_request_internal`.  In this case, we don't need to parse json twice, and we also don't need to find needed fields manually.
Could you please have a look?